presto-main/src/main/java/io/prestosql/operator/aggregation/ApproximateMostFrequentHistogram.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.aggregation;
+
+import com.clearspring.analytics.stream.Counter;
+import com.clearspring.analytics.stream.StreamSummary;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ *  Calculate the histogram approximately for topk elements based on the
+ *  <i>Space-Saving</i> algorithm and the <i>Stream-Summary</i> data structure
+ *  as described in:
+ *  <i>Efficient Computation of Frequent and Top-k Elements in Data Streams</i>
+ *  by Metwally, Agrawal, and Abbadi
+ * @param <K>
+ */
+public class ApproximateMostFrequentHistogram<K>
+{
+    private static final byte FORMAT_TAG = 0;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(ApproximateMostFrequentHistogram.class).instanceSize();
+    // Larger capacity for stream summary improves the accuracy.
+    private static final int MAX_CAPACITY_FACTOR = 5;
+
+    private StreamSummary<K> streamSummary;
```

Comment:
The implementation of StreamSummary is very "object-y" and will end up causing problems with object allocations at some point. It's fine to use for now, but we should consider reimplementing using flat structures (see, for instance, the implementation of QuantileDigest)

Reply:
Make this `final`